### PR TITLE
Rds/kjw/rule 18 1

### PR DESCRIPTION
**Pull request by KarenWGard**
please take a look - this is my first rule, so bound to be some issues

### CODE REVIEW

File: docs/section18/Rule_18-1 [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+2. get the list of HVAC systems and zones: `list_of_systems_and_zones = get_HVAC_building_area_types_and_zones(B-RMR)`
+	- find the predominant HVAC BAT by first sorting the list by area: `sorted_list = sorted(list_of_systems_and_zones.items(), key=lambda x: x[1]["AREA"], reverse = True)`
+	- the predominant building area type will be the first item in the sorted list: `if sorted_list[0][0] == "PUBLIC_ASSEMBLY":`
+		- set area_type_applicable to true: `area_type_applicable = TRUE`
+		
+3. find the building area and check whether it's less than 120,000 ft2.  Start by creating an area variable: `building_area = 0`
+	- We can use the list_of_systems_and_zones to get the total building floor area.  Start by looping through the list: `for x in list_of_systems_and_zones:`
+		- x will be the various HVAC BAT.  list_of_systems_and_zones[x]["AREA"] will give the total area of this BAT: `building_area = building_area + list_of_systems_and_zones[x]["AREA"]
+	- check if building_area < 120000: `if building_area < 120000:`
+		- set area_applicable to true: `area_applicable = TRUE`
```
Comment (mkarpman) [inline docs/section18/Rule_18-1:42]: We agreed that HVAC system type is set based on the total building area, not just the area for the given BAT, right?

File: docs/section18/Rule_18-1 [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+2. predominant HVAC BAT = public assembly
+3. building area < 120,0000 ft2
+4. the zone does not meet one if the G3.1.1.1 exceptions
+
+**Manual Check:** Yes  
+**Evaluation Context:** Evaluate whether each Zone is served by the correct HVAC system type  
+**Data Lookup:**   
+**Function Call:** 
+
+1. zone_meets_any_G3_1_1_exception(P-RMR,B-RMR,zone)
```
Comment (mkarpman) [inline docs/section18/Rule_18-1:23]: We don't yet have these function, right? Perhaps it makes sense to develop it before using it, so that we know the logic and returned variables. You defined function [zone_meets_G3_1_1c](https://github.com/pnnl/ruleset-checking-tool/pull/683), but this particular exception does not apply to System 4 that is checked below.

File: docs/section18/Rule_18-1 [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	- We can use the list_of_systems_and_zones to get the total building floor area.  Start by looping through the list: `for x in list_of_systems_and_zones:`
+		- x will be the various HVAC BAT.  list_of_systems_and_zones[x]["AREA"] will give the total area of this BAT: `building_area = building_area + list_of_systems_and_zones[x]["AREA"]
+	- check if building_area < 120000: `if building_area < 120000:`
+		- set area_applicable to true: `area_applicable = TRUE` 
+
+
+4. make a list of eligible zones: `eligible_zones = []`
+- loop through building segments: `for segment in RMR.building_segments:`
+	- loop through zones in the segments: `for zone in segment.zones:`
+		- check whether the zone meets one of the exceptions: `if(zone_meets_any_G3_1_1_exception(P-RMR,B-RMR,zone) == FALSE:`
```
Comment (mkarpman) [inline docs/section18/Rule_18-1:48]: I think it would be good to define this function before using it. Or is it already defined? Sorry if I missed it.

File: docs/section18/Rule_18-1 [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			- zone should take the predominant system type: `zone_applicable = TRUE`
+			- add zone to list: `eligible_zones.append zone`
+
+- if the 3 first checks are TRUE, the rule applies else, rule is not applicable to B-RMR: `else: RULE_NOT_APPLICABLE`
+
+## Rule Logic:  
+- call get_baseline_system_types and save a list of baseline HVAC system types: `hvac_system_types = get_baseline_system_types(B-RMR)`
+- for each zone that passed the zone applicability test above: `for zone in eligible_zones:`
```
Comment (mkarpman) [inline docs/section18/Rule_18-1:56]: The appicability checks shoudl be expressed in pseudo code withing the rule logic.

Comment (mkarpman) [reply docs/section18/Rule_18-1:56]: We already have list of HVAC systems associated with each zone - [get_list_hvac_systems_associated_with_zone.md](https://github.com/pnnl/ruleset-checking-tool/blob/develop/docs/ruleset_functions/get_list_hvac_systems_associated_with_zone.md)
So I think you just need to call this function for the "eligible_zone" and confirm that the zone has only one HVAC system and that system is of type 4.

File: docs/section18/Rule_18-1 [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# HVAC_SystemZoneAssignment - Rule 18-1  
+**Schema Version:** 0.0.22  
+
+**Rule ID:** 18-1  
+**Rule Description:** Buildings in CZ 0-3A with predominant HVAC BAT = public assembly, <120,000 ft2 have baseline HVAC system type 4 (PSZ HP)
+**Rule Assertion:** B-RMR = expected value  
+**Appendix G Section:** Section 18 HVAC_SystemZoneAssignment  
+**Appendix G Section Reference:** Table G3.1.1.3
+
```
[the remark this answers is not in the document]
Comment (weilixu) [reply docs/section18/Rule_18-1:9]: The function `get_HVAC_building_area_types_and_zones` in this PR provides building area to zone total area and zone id list.

`{"OTHER_NON_RESIDENTIAL": {"AREA": 13,000, "ZONE_LIST": ["zone_1", "zone_2"]}`

Would that work? It seems it is the same as the `get_zone_target_baseline_system_type`, but just the opposite mapping direction.

File: docs/section18/Rule_18-1 [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+4. make a list of eligible zones: `eligible_zones = []`
+- loop through building segments: `for segment in RMR.building_segments:`
+	- loop through zones in the segments: `for zone in segment.zones:`
+		- check whether the zone meets one of the exceptions: `if(zone_meets_any_G3_1_1_exception(P-RMR,B-RMR,zone) == FALSE:`
+			- zone should take the predominant system type: `zone_applicable = TRUE`
+			- add zone to list: `eligible_zones.append zone`
+
+- if the 3 first checks are TRUE, the rule applies else, rule is not applicable to B-RMR: `else: RULE_NOT_APPLICABLE`
```
Comment (weilixu) [inline docs/section18/Rule_18-1:52]: Seems there are four checks?

File: docs/section18/Rule_18-1 [hunk 7]
```diff
@@ -0,0 +1,72 @@
+# HVAC_SystemZoneAssignment - Rule 18-1  
+**Schema Version:** 0.0.22  
+
```
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1]: rules are here:

https://onedrive.live.com/Edit.aspx?resid=796769604E905084!199&wde=xlsx&wdorigin=Other&wdredirectionreason=Force_SingleStepBoot&wdinitialsession=34f6596c-f4e2-47b9-b433-bb11b1580e35&wdrldsc=1&wdrldc=3&wdrldr=FileOpenUserUnauthorized

formatting has been updated

File: docs/section18/Rule_18-1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1]: the master function has been created, and rule logic changed completely

File: docs/section18/Rule_18-1.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+2. baseline_system_type_compare()
+3. get_list_hvac_systems_associated_with_zone()
+4. get_baseline_system_types()
+5. is_zone_a_vestibule()
+
+
+**Applicability Checks:**
+- call the function get_zone_target_baseline_system(): `target_baseline_system_dict = get_zone_target_baseline_system(P_RMI, B_RMI)`
+- if the zone is conditioned and not dedicated to parking, the rule is applicable - the zone will only be in the dictionary provided by get_zone_target_baseline_system() if it is a zone which requires an HVAC system: `if zone in target_baseline_system_dict: continue to rule logic`
+- otherwise: `else: result = NOT_APPLICABLE`
```
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1.md:23]: the note here is lazy, but the filtering by conditioning is done in `get_zone_target_baseline_system`

File: docs/section18/Rule_18-1.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ - check if the system_type_origin is "G3_1_1d" we need to determine whether the laboratory exhaust meets the threshold based solely on zone exhaust or not: `if system_type_origin == "G3_1_1d:"
+	- complete logic to determine whether we can give a hard PASS / FAIL or simply UNDETERMINED
+
+ - - Now find the laboratory exhaust for each zone in the laboratory_zones list.  Loop through each zone: `for z in laboratory_zones:`
+   - Add zonal exhaust fan airflow to zone_total_exhaust: `if zone.zone_exhaust_fans ?:`
+       - look at each exhaust fan: `for exhaust_fan in zone.zone_exhaust_fans:`
+       -  add the airflow to the zone total exhaust: `building_total_lab_exhaust += exhaust_fan.design_airflow`
+       -  if the building total lab exhaust is greater than 15,000cfm, then the system selection is solid and we can give a hard PASS or FAIL, but if it is less than 15,000cfm, the system selection was based in part on HVAC system exhaust and we can't guarantee that it's all laboratory exhaust.  Check if the building_total_lab_exhaust is less than or equal to 15,000cfm: `if building_total_lab_exhaust <= 15,000:`
+		-  change the result to UNDETERMINED: `result = UNDETERMINED`	-
```
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1.md:46]: discussed on the call 6/22/2023 - looks ok.

File: docs/section18/Rule_18-1.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+- set the system_type_origin to the SYSTEM_ORIGIN given by the target_baseline_system_dict: `system_type_origin = target_baseline_system_type_dict[zone]["SYSTEM_ORIGIN"]
+- get the baseline system types in the building: `baseline_hvac_system_dict = get_baseline_system_types(B-RMI)`
+- now check whether the system(s) serving this zone match the expected system.  Start by getting the list of HVAC systems that serve the zone: `hvac_systems_serving_zone = get_list_hvac_systems_associated_with_zone(B-RMI)`
+- get a list of all of the HVAC systems that are of the same type as the expected system type for the zone: `systems_of_expected_type_list = baseline_hvac_system_dict[expected_system_type]`
+- loop through the systems that serve the zone: `for system_b in hvac_systems_serving_zone:`
+	- check to see if system_b is in the systems_of_expected_type_list, set result to PASS: `if system_b in systems_of_expected_type_list: result = PASS`
+ - create a variable advisory_note, which will give the user some feedback about how the HVAC system type was selected: `advisory_note = "HVAC system type " + expected_system_type + " was selected based on " + system_type_origin
```
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1.md:38]: changed logic so that it gives PASS only if all the systems assigned to the zone are the expected_system_type

[the remark this answers is not in the document]
Comment (mkarpman) [reply docs/section18/Rule_18-1.md:39]: The logic is a little hard to follow. Should it just be that if all HVAC systems in hvac_systems_serving_zone are the same type as the expected_system_type then PASS, else FAIL?

[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1.md:38]: changed

File: docs/section18/Rule_18-1.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (KarenWGard) [reply docs/section18/Rule_18-1.md:24]: changed

**Review by mkarpman — APPROVED** [review]
